[iOS] Sometimes starting an video AirPlay session will result in audio-only AirPlay

jernoble opened the pull request:
#### cd39c790c56eee7b006041caa16f9631ba460f00
<pre>
[iOS] Sometimes starting an video AirPlay session will result in audio-only AirPlay
<a href="https://rdar.apple.com/146166461">rdar://146166461</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=289467">https://bugs.webkit.org/show_bug.cgi?id=289467</a>

Reviewed by Eric Carlson.

From code inspection, it appears that sometimes VideoPresentationInterfaceIOS may be using
stale RoutingPolicy or RemoteContextUUID information when overriding the policy and contextID
used by AVPlayerViewController. The policy and contextID are queried when the view controller
is created, and they may change during the lifetime of the view controller. Add a client
interface to HTMLMediaElement and an observer on AudioContext to notify all the way up to
the UIProcess when the AudioContext changes its category information. This will trigger
the VideoPresentationInterfaceIOS to re-query for the policy and contextID, which should
ensure those pieces of information stay in sync during the view controller&apos;s lifetime.

* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::addClient):
(WebCore::HTMLMediaElement::removeClient):
(WebCore::HTMLMediaElement::audioSessionCategoryChanged):
* Source/WebCore/html/HTMLMediaElement.h:
(WebCore::HTMLMediaElementClient::audioSessionCategoryChanged):
* Source/WebCore/html/MediaElementSession.cpp:
(WebCore::MediaElementSession::audioSessionCategoryChanged):
* Source/WebCore/html/MediaElementSession.h:
* Source/WebCore/platform/audio/PlatformMediaSession.h:
(WebCore::PlatformMediaSession::audioSessionCategoryChanged):
* Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm:
(WebCore::MediaSessionManagerCocoa::updateSessionState):
* Source/WebCore/platform/cocoa/VideoPresentationModel.h:
(WebCore::VideoPresentationModelClient::audioSessionCategoryChanged):
* Source/WebCore/platform/cocoa/VideoPresentationModelVideoElement.h:
* Source/WebCore/platform/cocoa/VideoPresentationModelVideoElement.mm:
(WebCore::VideoPresentationModelVideoElement::cleanVideoListeners):
(WebCore::VideoPresentationModelVideoElement::setVideoElement):
(WebCore::VideoPresentationModelVideoElement::audioSessionCategoryChanged):
* Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.h:
* Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.mm:
(WebCore::VideoPresentationInterfaceIOS::setPlayerIdentifier):
(WebCore::VideoPresentationInterfaceIOS::audioSessionCategoryChanged):
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.messages.in:
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
(WebKit::VideoPresentationModelContext::audioSessionCategoryChanged):
(WebKit::VideoPresentationManagerProxy::audioSessionCategoryChanged):
* Source/WebKit/WebProcess/cocoa/VideoPresentationManager.h:
* Source/WebKit/WebProcess/cocoa/VideoPresentationManager.mm:
(WebKit::VideoPresentationInterfaceContext::audioSessionCategoryChanged):
(WebKit::VideoPresentationManager::audioSessionCategoryChanged):

Canonical link: <a href="https://commits.webkit.org/291912@main">https://commits.webkit.org/291912@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4919a8f72d6e0300d207465b1fb3f1d5ecfceb0f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/94341 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/13928 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3701 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/99353 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/44869 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/96391 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/14229 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/22358 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/72005 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29337 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/97343 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/10874 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/85201 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52327 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/10278 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/2881 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/44187 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/80511 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/2982 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101398 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21393 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/15620 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/81008 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21645 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/81207 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80366 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24921 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/2303 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/14622 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15144 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21377 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/26563 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/21064 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24524 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22805 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.mm:
```diff
@@ -249,6 +249,21 @@ @interface UIViewController ()
 
+void VideoPresentationInterfaceIOS::audioSessionCategoryChanged(WebCore::AudioSessionCategory, WebCore::AudioSessionMode, WebCore::RouteSharingPolicy)
+{
+    auto model = videoPresentationModel();
```
`auto` -> `RefPtr` as I understand we don't want to hide smart pointers behind auto in new code

jernoble committed:
[iOS] Sometimes starting an video AirPlay session will result in audio-only AirPlay
rdar://146166461
https://bugs.webkit.org/show_bug.cgi?id=289467

Reviewed by Eric Carlson.

From code inspection, it appears that sometimes VideoPresentationInterfaceIOS may be using
stale RoutingPolicy or RemoteContextUUID information when overriding the policy and contextID
used by AVPlayerViewController. The policy and contextID are queried when the view controller
is created, and they may change during the lifetime of the view controller. Add a client
interface to HTMLMediaElement and an observer on AudioContext to notify all the way up to
the UIProcess when the AudioContext changes its category information. This will trigger
the VideoPresentationInterfaceIOS to re-query for the policy and contextID, which should
ensure those pieces of information stay in sync during the view controller's lifetime.

* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::addClient):
(WebCore::HTMLMediaElement::removeClient):
(WebCore::HTMLMediaElement::audioSessionCategoryChanged):
* Source/WebCore/html/HTMLMediaElement.h:
(WebCore::HTMLMediaElementClient::audioSessionCategoryChanged):
* Source/WebCore/html/MediaElementSession.cpp:
(WebCore::MediaElementSession::audioSessionCategoryChanged):
* Source/WebCore/html/MediaElementSession.h:
* Source/WebCore/platform/audio/PlatformMediaSession.h:
(WebCore::PlatformMediaSession::audioSessionCategoryChanged):
* Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm:
(WebCore::MediaSessionManagerCocoa::updateSessionState):
* Source/WebCore/platform/cocoa/VideoPresentationModel.h:
(WebCore::VideoPresentationModelClient::audioSessionCategoryChanged):
* Source/WebCore/platform/cocoa/VideoPresentationModelVideoElement.h:
* Source/WebCore/platform/cocoa/VideoPresentationModelVideoElement.mm:
(WebCore::VideoPresentationModelVideoElement::cleanVideoListeners):
(WebCore::VideoPresentationModelVideoElement::setVideoElement):
(WebCore::VideoPresentationModelVideoElement::audioSessionCategoryChanged):
* Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.h:
* Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.mm:
(WebCore::VideoPresentationInterfaceIOS::setPlayerIdentifier):
(WebCore::VideoPresentationInterfaceIOS::audioSessionCategoryChanged):
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.messages.in:
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
(WebKit::VideoPresentationModelContext::audioSessionCategoryChanged):
(WebKit::VideoPresentationManagerProxy::audioSessionCategoryChanged):
* Source/WebKit/WebProcess/cocoa/VideoPresentationManager.h:
* Source/WebKit/WebProcess/cocoa/VideoPresentationManager.mm:
(WebKit::VideoPresentationInterfaceContext::audioSessionCategoryChanged):
(WebKit::VideoPresentationManager::audioSessionCategoryChanged):

Canonical link: https://commits.webkit.org/291912@main
